keps/sig-node/3063-dynamic-resource-allocation/README.md

```diff
@@ -0,0 +1,1549 @@
+
+For a resource driver the following components are needed:
+- Some utility library similar to
+  https://github.com/kubernetes-sigs/sig-storage-lib-external-provisioner
+  and the code in driver-registrar.
+- *Resource driver controller*: a central component which handles resource allocation
+  by watching and modifying ResourceClaims.
+- *Resource kubelet plugin*: a component which cooperates with kubelet to prepare
+  the usage of the resource on a node.
+
+The utility library will be developed outside of Kubernetes and does not have
+to be used by drivers, therefore it is not described further in this KEP.
+
+### Resource allocation flow
+
+The following diagram shows how resource allocation works for a resource that
+gets defined inline in a Pod. For a full definition of ResourceClass,
+ResourceClaim and ResourceClaimTemplate see the [API](#API) section below.
+
+Several of these operations may fail without changing the system state. They
+then must be retried until they succeed or something else changes in the
+system, like for example deleting objects. These additional state transitions
+are not shown for the sake of simplicity.
+
+![allocation](./allocation.png)
+
+The flow is similar for a ResourceClaim that gets created as a stand-alone
+object by the user. In that case, the Pod reference that ResourceClaim by
+name. The ResourceClaim does not get deleted at the end and can be reused by
+another Pod and/or used by multiple different Pods at the same time (if
+supported by the driver).
+
+### API
+
+ResourceClaim, ResourceClass and ResourceDriver are new built-in types
+in a new `cdi.k8s.io/v1alpha1` API group. This was chosen instead of
+using CRDs because core Kubernetes components must interact with them
+and installation of CRDs as part of cluster creation is an unsolved
+problem.
+
+The PodSpec gets extended.
+
+Secrets are not part of this API: if a resource driver needs secrets, for
+example to access its own backplane, then it can define custom parameters for
+those secrets and retrieve them directly from the apiserver. This works because
+drivers are expected to be written for Kubernetes.
+
+```
+type ResourceClass struct {
+    // Resource drivers have a unique name in reverse domain order (acme.example.com).
+    DriverName string
+    // Parameters holds arbitrary values that will be available to the driver
+    // when allocating a resource that uses this class. The driver will
+    // be able to distinguish between parameters stored here and
+    // and those stored in ResourceClaimSpec. These parameters
+    // here can only be set by cluster administrators.
+    Parameters runtime.RawExtension
+}
+
+type ResourceDriver struct {
+    // The name of the object is the unique driver name.
+    ObjectMeta
+
+    // Features contains a list of features supported by the driver.
+    // New features may be added over time and must be ignored
+    // by code that does not know about them.
+    Features []ResourceDriverFeature
+}
```

Comment:
ResourceDriver currently doesn't have a real usage yet. I think it is useful to document how we intend to implement feature information, but perhaps we can take it out of the initial alpha?

Alternatively, what if we do make it useful and add some informational fields:
- `Description string` - short, single-line string
- `Help string` - longer, potentially multi-line string explaining what the ResourceDriver is
- `ClassParameterHelp string` - explanation of valid ResourceClass.Parameter values
- `ClaimParameterHelp string` - explanation of valid ResourceClaim.Parameter values

As an administrator and user I would find it useful to have a way to discover this information via kubectl, in particular because there is no other information about those parameter blobs.

Reply:
There is no such core component at the moment.

Let's leave the ResourceDriver in the KEP, but with a comment saying "this is how we could handle driver information if we needed to", and not implement this in v1alpha1.

Reply:
Move it to "Alternatives" section.

Reply:
Done.